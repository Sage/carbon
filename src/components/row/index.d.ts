export { default as Row } from "./row";
export { default as Column } from "./column/column";
