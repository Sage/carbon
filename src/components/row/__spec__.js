import React from 'react';
import Immutable from 'immutable';
import TestUtils from 'react/lib/ReactTestUtils';
import Row from './row';

describe('Row', () => {

  describe('when column number is NOT passed', () => {
    let instance;
    let columns;

    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Row className="foobar">
          <div columnOffset={2}>Foo</div>
          <div columnSpan={3}>Bar</div>
          { null }
          <div columnClasses='extra-class'>Bar</div>
          <div columnAlign='center'>Bar</div>
        </Row>
      );

      columns = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-row__column');
    });

    describe('render', () => {
      it('renders a parent div with calculated CSS classes', () => {
        let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
        expect(rowNode.className).toEqual('carbon-row foobar carbon-row--columns-4');
      });

      it('renders the correct amount of columns', () => {
        expect(columns.length).toEqual(4);
      });
    });

    describe('with immutable data', () => {
      it('renders the correct number of columns', () => {
        let data = Immutable.fromJS([{
          name: 'foo'
        }, {
          name: 'bar'
        }]);

        instance = TestUtils.renderIntoDocument(
          <Row className="foobar">
            {
              data.map((item) => {
                return <div>{ item.get('name') }</div>;
              })
            }
          </Row>
        );

        columns = TestUtils.scryRenderedDOMComponentsWithClass(instance, 'carbon-row__column');
        expect(columns.length).toEqual(2);
      });
    });

    describe('with no children', () => {
      it('allows render', () => {
        instance = TestUtils.renderIntoDocument(
          <Row>{ null }</Row>
        );
        let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
        expect(rowNode).toBeTruthy();
      });
    });

    describe('with no children in an array', () => {
      it('allows render', () => {
        let children = [];
        instance = TestUtils.renderIntoDocument(
          <Row>{ children }</Row>
        );
        let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
        expect(rowNode).toBeTruthy();
      });
    });

    describe('Column offset', () => {
      it('renders a div with an additional offset CSS class', () => {
        expect(columns[0].className).toEqual('carbon-row__column carbon-row__column--offset-2');
      });
    });

    describe('Column span', () => {
      it('renders a div with an additional span CSS class', () => {
        expect(columns[1].className).toEqual('carbon-row__column carbon-row__column--span-3');
      });
    });

    describe('Column classes', () => {
      it('renders a div with all additional column classes', () => {
        expect(columns[2].className).toEqual('carbon-row__column extra-class');
      });
    });

    describe('Column align', () => {
      it('renders a div with alignment class', () => {
        expect(columns[3].className).toEqual('carbon-row__column carbon-row__column--align-center');
      });
    });
  });

  describe('When column number is passed', () => {
    let instance;

    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Row columns={2}>
          <div>Foo</div>
          <div>Bar</div>
        </Row>
      );
    });

    it('renders a parent div with calculated CSS classes', () => {
      let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
      expect(rowNode.className).toEqual('carbon-row carbon-row--columns-2');
    });
  });

  describe('When there is only 1 column', () => {
    let instance;

    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Row>
          <div>Foo</div>
        </Row>
      );
    });

    it('renders a parent div with calculated CSS classes', () => {
      let rowNode = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-row')
      expect(rowNode.className).toEqual('carbon-row carbon-row--columns-1');
    });
  });
});
