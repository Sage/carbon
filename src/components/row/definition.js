import Row from './';
import OptionsHelper from 'utils/helpers/options-helper';
import ColumnDefinition from './column/definition';
import Definition from './../../../demo/utils/definition';

let definition = new Definition('row', Row, {
  description: `Sets up a basic column-based UI layout.`,
  associatedDefinitions: [ColumnDefinition],
  designerNotes: `
* Useful to organise the UI of a page into a simple column-based layout.
* Configure the number of columns, the margin between them, and any separators.
  `,
  relatedComponentsNotes: `
* Need an overall container? [Try App Wrapper](/components/app-wrapper).
* Need a container for your primary navigation? [Try Navigation Bar](/components/navigation-bar).
* Need a layout with controls and guidance text? [Try Settings Row](/components/settings-row).
 `,
  hiddenProps: ['children'],
  propTypes: {
    gutter: "String",
    columnDivide: "Boolean",
    columns: "String"
  },
  propDescriptions: {
    gutter: "Define how wide the gutter between the rows and columns should be.",
    columnDivide: "Enable a divider between each column.",
    columns: "Define a certain amount of columns, instead of basing it on the number of children.",
    children: "This component supports children of type Column."
  },
  propOptions: {
    gutter: OptionsHelper.sizesFull
  }
});

definition.addChildByDefinition(ColumnDefinition, { children: `<div />` });
definition.addChildByDefinition(ColumnDefinition, { children: `<div />` });
definition.addChildByDefinition(ColumnDefinition, { children: `<div />` });
definition.addChildByDefinition(ColumnDefinition, { children: `<div />` });
definition.addChildByDefinition(ColumnDefinition, { children: `<div />` });
definition.addChildByDefinition(ColumnDefinition, { children: `<div />` });

export default definition;
