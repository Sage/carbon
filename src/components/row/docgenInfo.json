{
  "src/components/row/column/column.js": [
    {
      "description": "",
      "displayName": "Column",
      "methods": [],
      "props": {
        "children": {
          "type": {
            "name": "node"
          },
          "required": false,
          "description": "This component supports children."
        },
        "className": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Classes to apply to the component."
        },
        "columnClasses": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Classes applied by row component to affect all rows"
        },
        "columnDivide": {
          "type": {
            "name": "bool"
          },
          "required": false,
          "description": "Show a divide between columns. This is defined by the row component."
        },
        "columnAlign": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Alignment of content within column.",
          "defaultValue": {
            "value": "'left'",
            "computed": false
          }
        },
        "columnOffset": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Offset this column by a certain number of columns.",
          "defaultValue": {
            "value": "'0'",
            "computed": false
          }
        },
        "columnSpan": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Span this column by a certain number of columns.",
          "defaultValue": {
            "value": "'1'",
            "computed": false
          }
        }
      }
    }
  ],
  "src/components/row/row.js": [
    {
      "description": "",
      "displayName": "Row",
      "methods": [
        {
          "name": "buildColumns",
          "docblock": "Builds row columns from the children object fields\n\n@method buildColumns\n@return {Array} array of built columns",
          "modifiers": [],
          "params": [],
          "returns": {
            "description": "array of built columns",
            "type": {
              "name": "Array"
            }
          },
          "description": "Builds row columns from the children object fields"
        },
        {
          "name": "mainClasses",
          "docblock": "Main Class getter\n\n@method mainClasses\n@return {String} Main className",
          "modifiers": [
            "get"
          ],
          "params": [],
          "returns": {
            "description": "Main className",
            "type": {
              "name": "String"
            }
          },
          "description": "Main Class getter"
        }
      ],
      "props": {
        "children": {
          "type": {
            "name": "union",
            "value": [
              {
                "name": "array"
              },
              {
                "name": "object"
              }
            ]
          },
          "required": false,
          "description": "This component supports children of type Column."
        },
        "className": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Classes to apply to the component."
        },
        "gutter": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Define how wide the gutter between the rows and columns should be.",
          "defaultValue": {
            "value": "'medium'",
            "computed": false
          }
        },
        "columnDivide": {
          "type": {
            "name": "bool"
          },
          "required": false,
          "description": "Enable a divider between each column."
        },
        "columns": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Define a certain amount of columns, instead of basing it on the number of children."
        },
        "columnClasses": {
          "type": {
            "name": "string"
          },
          "required": false,
          "description": "Classes to apply to all column children."
        }
      }
    }
  ]
}