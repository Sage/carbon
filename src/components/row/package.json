{
  "name": "row",
  "style": "row.scss",
  "main": "./row.js"
}
