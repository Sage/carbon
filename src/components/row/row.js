import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import Column from './column';

/**
 * A row widget.
 *
 * This is a standalone row widget used for layout; for table rows use the table-row widget.
 *
 * == How to use a Row in a component:
 *
 * In your file
 *
 *   import { Row, Column } from 'carbon-react/lib/components/row';
 *
 * To render the Row:
 *
 *   <Row>
 *     <Column>Column1</Column>
 *     <Column>Column2</Column>
 *   </Row>
 *
 * A Rows child must be of type Column
 *
 * @class Row
 * @constructor
 */
class Row extends React.Component {
  static propTypes = {

    /**
     * The elements to be rendered in the row
     *
     * @property children
     * @type {Object | Array}
     */
    children: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.object
    ]),

    /**
     * Custom className
     *
     * @property className
     * @type {String}
     */
    className: PropTypes.string,

    /**
     * Pass a custom value for the gutter
     * (extra-small, small, medium, large or extra-large)
     *
     * @property gutter
     * @type {String}
     */
    gutter: PropTypes.string,

    /**
     * Show a divide between columns
     *
     * @property columnDivide
     * @type {String}
     */
    columnDivide: PropTypes.bool,

    /**
     * Manually define number of columns
     *
     * @property columns
     * @type {String}
     */
    columns: PropTypes.string,

    /**
     * class to apply to each child column
     *
     * @property columnClasses
     * @type {String}
     */
    columnClasses: PropTypes.string
  }

  static defaultProps = {
    gutter: 'medium'
  };

  /**
   * Builds row columns from the children object fields
   *
   * @method buildColumns
   * @return {Array} array of built columns
   */
  buildColumns = () => {
    return React.Children.toArray(this.props.children).map((child) => {
      return React.cloneElement(
        child, {
          columnClasses: this.props.columnClasses,
          columnDivide: this.props.columnDivide
        },
        child.props.children
      );
    });
  }

  /**
   * Main Class getter
   *
   * @method mainClasses
   * @return {String} Main className
   */
  get mainClasses() {
    const columns = this.props.columns || React.Children.toArray(this.props.children).length;

    return classNames(
      'carbon-row',
      `carbon-row--gutter-${this.props.gutter}`,
      this.props.className,
      `carbon-row--columns-${columns}`
    );
  }

  /**
   * Renders the component.
   *
   * @method render
   * @return {Object} JSX
   */
  render() {
    return (
      <div className={ this.mainClasses }>
        { this.buildColumns() }
      </div>
    );
  }
}

export default Row;
export {
  Row,
  Column
};
