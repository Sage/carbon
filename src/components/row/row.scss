@import 'general';

$max-columns: 12 !default;
$row-margin: 10px !default;

.ui-row {
  clear: both;
  margin-bottom: -($row-margin);
  margin-left: -($grid-margin);
  position: relative;

  + .ui-row {
    margin-top: $row-margin;
  }

  .ui-row__column {
    box-sizing: border-box;
    float: left;
    margin-bottom: 10px;
    min-height: 1px;
    padding-left: $grid-margin;
    width: 100%;
  }

  @for $i from 2 through $max-columns {
    $width: 100% / $i;

    &.ui-row--columns-#{$i} {
      > .ui-row__column {
        width: $width;
      }

      // span columns:
      @for $y from 2 through $i {
        @if $y != $i {
          > .ui-row__column--span-#{$y} {
            width: (100% / $i) * $y;
          }
        }
      }

      // offset columns:
      @for $y from 1 through $i {
        @if $y != $i {
          > .ui-row__column--offset-#{$y} {
            margin-left: $width * $y;
          }
        }
      }
    }
  }

  // Clearfix:
  &:before,
  &:after {
    content: " ";
    display: table;
  }
  &:after {
    clear: both;
  }
}
