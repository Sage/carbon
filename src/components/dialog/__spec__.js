import React from 'react';
import TestUtils from 'react-dom/test-utils';
import Bowser from 'bowser';
import { shallow, mount } from 'enzyme';
import Browser from '../../utils/helpers/browser';
import Dialog from './dialog';
import Button from './../button';
import { Row, Column } from './../row';
import { elementsTagTest, rootTagTest } from '../../utils/helpers/tags/tags-specs';

describe('Dialog', () => {
  let instance, onCancel;

  beforeEach(() => {
    onCancel = jasmine.createSpy('cancel');
  });

  describe('Lifecycle functions', () => {
    describe('componentDidMount', () => {
      describe('when dialog is open', () => {
        it('centers the dialog', () => {
          instance = TestUtils.renderIntoDocument(
            <Dialog open={ true } onCancel={ onCancel } />
          );
          spyOn(instance, "centerDialog");
          instance.componentDidMount();
          expect(instance.centerDialog).toHaveBeenCalled();
        });

        it('focuses on the dialog', () => {
          spyOn(Dialog.prototype, 'focusDialog');
          const wrapper = mount(
            <Dialog open={ true } onCancel={ onCancel } />
          );
          expect(Dialog.prototype.focusDialog).toHaveBeenCalled();
        });
      });

      describe('when dialog is closed', () => {
        it('does not center the dialog', () => {
          instance = TestUtils.renderIntoDocument(
            <Dialog open={ false } onCancel={ onCancel } />
          );
          spyOn(instance, "centerDialog");
          instance.componentDidMount();
          expect(instance.centerDialog).not.toHaveBeenCalled();
        });
      });
    });

    describe('componentDidUpdate', () => {
      let mockWindow;
      let wrapper;
      let instance;

      beforeEach(() => {
        mockWindow = {
          addEventListener() {},
          removeEventListener() {}
        };

        spyOn(Browser, 'getWindow').and.returnValue(mockWindow);
      });

      describe('when the dialog is open', () => {
        beforeEach(() => {
          wrapper = mount(
            <Dialog open={ true } onCancel={ onCancel } />
          );
          instance = wrapper.instance();
        });

        it('centers the dialog', () => {
          spyOn(instance, 'centerDialog');
          wrapper.setProps({ title: 'Dialog title' });
          expect(instance.centerDialog).toHaveBeenCalled();
        });

        it('sets up event listeners to resize and close the dialog', () => {
          spyOn(mockWindow, 'addEventListener');

          wrapper.setProps({ title: 'Dialog title' });
          expect(mockWindow.addEventListener.calls.count()).toEqual(2);
          expect(mockWindow.addEventListener).toHaveBeenCalledWith('resize', instance.centerDialog);
          expect(mockWindow.addEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
        });

        describe('when the dialog is already listening', () => {
          it('does not set up event listeners', () => {
            spyOn(mockWindow, 'addEventListener');

            instance.listening = true;
            wrapper.setProps({ title: 'Dialog title' });

            expect(mockWindow.addEventListener.calls.count()).toEqual(0);
            expect(mockWindow.addEventListener).not.toHaveBeenCalled();
            expect(mockWindow.addEventListener).not.toHaveBeenCalled();
          });
        });
      });

      describe('when the dialog is closed', () => {
        beforeEach(() => {
          wrapper = mount(
            <Dialog open={ false } onCancel={ onCancel } />
          );
          instance = wrapper.instance();
        });

        it('removes event listeners for resize and closing', () => {
          spyOn(mockWindow, 'removeEventListener');
          wrapper.setProps({ title: 'Dialog closed' });

          expect(mockWindow.removeEventListener.calls.count()).toEqual(2);
          expect(mockWindow.removeEventListener).toHaveBeenCalledWith('resize', instance.centerDialog);
          expect(mockWindow.removeEventListener).toHaveBeenCalledWith('keyup', instance.closeModal);
        });
      });
    });
  });

  describe('centerDialog', () => {
    beforeEach(() => {
      instance = TestUtils.renderIntoDocument(
        <Dialog open={ true } onCancel={ onCancel } />
      );
    });

    describe('when dialog is lower than 20px', () => {
      it('sets top position to the correct value', () => {
        instance.centerDialog();
        expect(instance._dialog.style.top).toEqual('150px');
      });
    });

    describe('when dialog is higher than 20px', () => {
      it('sets top position to 20px', () => {
        instance._dialog = {
          style: {},
          offsetHeight: 261
        };
        instance.centerDialog();
        expect(instance._dialog.style.top).toEqual('20px');
      });
    });

    describe('when dialog is less than 20px from the side', () => {
      it('sets top position to 20px', () => {
        instance._dialog = {
          style: {},
          offsetWidth: 361
        };
        instance.centerDialog();
        expect(instance._dialog.style.left).toEqual('20px');
      });
    });

    describe('when ios', () => {
      it('does not remove page y offset', () => {
        Bowser.ios = true;
        instance.centerDialog();
        expect(instance._dialog.style.top).toEqual('150px');
      });
    });
  });

  describe('dialogTitle', () => {
    describe('when a props title is passed', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Dialog
            onCancel={ onCancel }
            open={ true }
            title="Dialog title"
          />
        );
      });

      it('sets a dialog header', () => {
        let header = TestUtils.findRenderedDOMComponentWithTag(instance, 'h2');
        expect(header.classList[0]).toEqual('carbon-dialog__title');
        expect(header.textContent).toEqual('Dialog title');
      });
    });

    describe('when a props object title is passed', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Dialog
            onCancel={ onCancel }
            open={ true }
            title={
              <Row>
                <Column>Row1</Column>
                <Column>Row2</Column>
              </Row>
            }
          />
        );
      });

      it('sets a dialog header', () => {
        let header = TestUtils.findRenderedDOMComponentWithTag(instance, 'h2');
        expect(header.classList[0]).toEqual('carbon-dialog__title');
        expect(header.textContent).toEqual('Row1Row2');
      });
    });

    describe('when a props title is not passed', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Dialog
            onCancel={ onCancel }
            open={ true }
          />
        );
      });

      it('defaults to null', () => {
        expect(instance.dialogTitle).toBeFalsy();
      });
    });
  });

  describe('dialogSubTitle', () => {
    describe('when a props subtitle is passed', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Dialog
            onCancel={ onCancel }
            open={ true }
            subtitle="My informative subtitle"
          />
        );
      });

      it('sets a dialog subtitle', () => {
        let subtitle = TestUtils.findRenderedDOMComponentWithTag(instance, 'p');
        expect(subtitle.classList[0]).toEqual('carbon-dialog__subtitle');
        expect(subtitle.textContent).toEqual('My informative subtitle');
      });
    });

    describe('when a props subtitle is not passed', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Dialog
            onCancel={ onCancel }
            open={ true }
          />
        );
      });

      it('defaults to null', () => {
        expect(instance.dialogSubTitle).toBeFalsy();
      });
    });
  });

  describe('render', () => {
    describe('when dialog is open', () => {
      beforeEach(() => {
        instance = TestUtils.renderIntoDocument(
          <Dialog
            onCancel={ onCancel }
            className="foo"
            open={ true } >

            <Button>Button</Button>
            <Button>Button</Button>
          </Dialog>
        );
      });

      it('renders a parent div with mainClasses attached', () => {
        let dialogNode = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'div')[0];
        expect(dialogNode.className).toEqual('carbon-dialog foo');
      });

      it('renders the dialog', () => {
        expect(instance._dialog).toBeTruthy();
        expect(instance._dialog.classList[0]).toEqual('carbon-dialog__dialog');
      });

      it('closes when the exit icon is click', () => {
        let closeIcon = TestUtils.findRenderedDOMComponentWithClass(instance, 'carbon-dialog__close');
        TestUtils.Simulate.click(closeIcon);
        expect(onCancel).toHaveBeenCalled();
      });

      it('renders the children passed to it', () => {
        let buttons = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'button');
        expect(buttons.length).toEqual(2);
      });

      describe('when passing a custom size', () => {
        it('adds the size class to the dialog', () => {
          instance = TestUtils.renderIntoDocument(
            <Dialog
              onCancel={ onCancel }
              open={ true }
              size='small' />
          );

          expect(instance._dialog.classList[1]).toEqual('carbon-dialog__dialog--small');
        });
      });
    });

    describe('when dialog is closed', () => {
      it('renders a parent div with mainClasses attached', () => {
        instance = TestUtils.renderIntoDocument(
          <Dialog open={ false } onCancel={ onCancel } />
        );
        let dialogNode = TestUtils.scryRenderedDOMComponentsWithTag(instance, 'div')[0];
        expect(dialogNode.classList[0]).toEqual('carbon-dialog');
      });
    });
  });

  describe("tags", () => {
    describe("on component", () => {
      let wrapper = shallow(
        <Dialog
          data-element='bar'
          onCancel={ () => {} }
          onConfirm={ () => {} }
          open={ true }
          data-role='baz'
        />);

      it('include correct component, element and role data tags', () => {
        rootTagTest(wrapper, 'dialog', 'bar', 'baz');
      });
    });

    describe("on internal elements", () => {
      let wrapper = shallow(
        <Dialog
          onCancel={ () => {} }
          onConfirm={ () => {} }
          open={ true }
          showCloseIcon={ true }
          subtitle='Test'
          title='Test'
        />
      );

      elementsTagTest(wrapper, [
        'close',
        'subtitle',
        'title'
      ]);
    });
  });

  describe('a11y', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = mount(
        <Dialog
          onCancel={ () => {} }
          onConfirm={ () => {} }
          open={ true }
          showCloseIcon={ true }
          subtitle='Test'
          title='Test'
          ariaRole="dialog"
        />
      );
    });

    describe('when title, subtitle, and ariaRole are set', () => {
      it('renders a role attribute from the ariaRole prop', () => {
        expect(wrapper.find('[role="dialog"]').exists()).toBe(true);
      });

      it('renders an aria-labelledby pointing at the title element', () => {
        expect(wrapper.find('[aria-labelledby="carbon-dialog-title"]').exists()).toBe(true);
      });

      it('renders an aria-describedby attribute pointing at the subtitle element', () => {
        expect(wrapper.find('[aria-describedby="carbon-dialog-subtitle"]').exists()).toBe(true);
      });
    });

    describe('when title, subtitle, and ariaRole are not set', () => {
      beforeEach(() => {
        wrapper = mount(
          <Dialog
            onCancel={ () => {} }
            onConfirm={ () => {} }
            open={ true }
            showCloseIcon={ true }
            ariaRole=""
          />
        );
      });

      it('renders a role attribute from the ariaRole prop', () => {
        expect(wrapper.find('[role="dialog"]').exists()).toBe(false);
      });

      it('renders an aria-labelledby pointing at the title element', () => {
        expect(wrapper.find('[aria-labelledby="carbon-dialog-title"]').exists()).toBe(false);
      });

      it('renders an aria-describedby attribute pointing at the subtitle element', () => {
        expect(wrapper.find('[aria-describedby="carbon-dialog-subtitle"]').exists()).toBe(false);
      });
    });

    it('focuses on the dialog when opened', () => {
      wrapper.setProps({
        open: false
      });
      const instance = wrapper.instance();
      spyOn(instance, 'focusDialog');

      wrapper.setProps({
        open: true
      });
      expect(instance.focusDialog).toHaveBeenCalled();
    });

    it('returns focus to the dialog element when focus leaves the close icon', () => {
      const dialogElement = wrapper.find('[role="dialog"]').first().getDOMNode();
      spyOn(dialogElement, 'focus');

      const closeIcon = wrapper.find('[data-element="close"]');
      closeIcon.simulate('blur');
      expect(dialogElement.focus).toHaveBeenCalled();
    });
  });
});
