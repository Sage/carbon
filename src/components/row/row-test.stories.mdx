import { Meta, Story, Canvas } from "@storybook/addon-docs";

import specialCharacters from "../../../.storybook/utils/argTypes/specialCharacters";
import { ROW_SIZES } from "./row.config";
import { Row, Column } from ".";

<Meta
  title="Row/Test"
  parameters={{
    info: { disable: true },
    chromatic: {
      disable: true,
    },
  }}
  argTypes={{
    gutter: {
      options: ROW_SIZES,
      control: {
        type: "select",
      },
    },
    columnAlign: {
      options: ["left", "middle", "center", "right"],
      control: {
        type: "select",
      },
    },
    childrenSpecialCharacters: specialCharacters,
  }}
/>

export const RowStory = ({
  columnDivide,
  gutter,
  children,
  childrenSpecialCharacters,
  ...args
}) => (
  <Row columnDivide={columnDivide} gutter={gutter}>
    <Column children={children || childrenSpecialCharacters} {...args} />
    <Column children={children || childrenSpecialCharacters} {...args} />
    <Column children={children || childrenSpecialCharacters} {...args} />
    <Column children={children || childrenSpecialCharacters} {...args} />
  </Row>
);

# Row

### Default

<Canvas>
  <Story
    name="default"
    args={{
      columnDivide: true,
      gutter: Row.defaultProps.gutter,
      columnAlign: Column.defaultProps.columnAlign,
      columnOffset: Column.defaultProps.columnOffset,
      columnSpan: Column.defaultProps.columnSpan,
      children: "content",
      childrenSpecialCharacters: undefined,
    }}
  >
    {RowStory.bind({})}
  </Story>
</Canvas>
