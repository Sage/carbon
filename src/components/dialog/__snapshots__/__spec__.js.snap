// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog dialog headers when a props object title is passed wraps it within a div 1`] = `
<div
  className="carbon-dialog__title"
>
  <Row
    gutter="medium"
  >
    <Column>
      Row1
    </Column>
    <Column>
      Row2
    </Column>
  </Row>
</div>
`;

exports[`Dialog render when dialog is open has the correct content, tags, elements etc 1`] = `
<Portal
  key="1"
>
  <div
    className="carbon-dialog foo"
    data-component="dialog"
    data-element="bar"
    data-role="baz"
    data-state="open"
  >
    <CSSTransitionGroup
      component="div"
      transitionAppear={true}
      transitionAppearTimeout={500}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal-background"
    >
      <div
        className="carbon-modal__background"
      />
    </CSSTransitionGroup>
    <CSSTransitionGroup
      component="div"
      transitionAppear={true}
      transitionAppearTimeout={500}
      transitionEnter={true}
      transitionEnterTimeout={500}
      transitionLeave={true}
      transitionLeaveTimeout={500}
      transitionName="modal"
    >
      <div
        aria-describedby="carbon-dialog-subtitle"
        aria-labelledby="carbon-dialog-title"
        className="carbon-dialog__dialog carbon-dialog__dialog--small carbon-dialog__dialog--has-height"
        data-component="dialog"
        data-element="bar"
        data-role="baz"
        onBlur={[Function]}
        role="dialog"
        style={
          Object {
            "minHeight": "500px",
          }
        }
        tabIndex={0}
      >
        <div
          className="carbon-dialog__title carbon-dialog__title--has-subheader"
        >
          <Heading
            divider={true}
            separator={false}
            subheader="Test"
            subtitleId="carbon-dialog-subtitle"
            title="Test"
            titleId="carbon-dialog-title"
          />
        </div>
        <div
          className="carbon-dialog__content"
        >
          <div
            className="carbon-dialog__inner-content"
            style={
              Object {
                "minHeight": "calc(500px - 40px)",
              }
            }
          >
            <Button
              as="secondary"
              disabled={false}
              size="medium"
              subtext=""
              theme="blue"
            >
              Button
            </Button>
            <Button
              as="secondary"
              disabled={false}
              size="medium"
              subtext=""
              theme="blue"
            >
              Button
            </Button>
          </div>
        </div>
        <Icon
          bgSize="small"
          className="carbon-dialog__close"
          data-element="close"
          onBlur={[Function]}
          onClick={[Function]}
          tabIndex="0"
          type="close"
        />
      </div>
    </CSSTransitionGroup>
  </div>
</Portal>
`;
